### PR TITLE
Dataloader: datamodule and datasets refactoring (1/3)

### PR DESCRIPTION
**Pull request by sfmig**
This PR changes the dataset and datamodule modules to make full use of the lightning pose options. 

### Features:
- **datamodules.py**: a module to host Lightning pose datamodule classes. These are meant to encapsulate all the steps needed to process the data. For us it includes: the dataset split into train and test, the transformations for train and test sets, and the definition of the dataloaders. 
    - I added the option of defining a validation split, but it is set to 0 for now.
    - I changed the Pytorch`transforms` to the `v2` version, following [this](https://pytorch.org/vision/main/auto_examples/transforms/plot_transforms_e2e.html#transforms) guide and [this](https://pytorch.org/vision/stable/transforms.html#v1-or-v2-which-one-should-i-use) recommendations.
    - The dataset split is reproducible if we pass the same `split_seed`

- **datasets.py**: this module is meant to host the definition of the dataset(s) as a map from keys to samples. Main additions:
    - We can pass multiple dataset directories and they are combined into one dataset.
    - We can pass a list of files to exclude from the dataset.
    - We are now reusing the existing `CocoDetection` class from Pytorch (before we were slightly re-implementing).
    - We use the `wrap_dataset_for_transforms_v2` to format the dataset for Pytorch's transforms.
    - I got a helper plotting function from Pytorch vision examples, to easily plot samples from the data. But maybe this lives better somewhere else?
       ```
       sample = dataset[0]  # tuple with (image, annotations)
	   plot_sample([sample])
	   ```
- **CLI inputs to train_model.py**: we can run training similarly to before. The only required argument is the list of dataset directories. If no annotation files are passed, they are assumed to be under the corresponding dataset, at `dataset/annotations/VIA_JSON_combined_coco_gen.json`. For the annotation files, we can also pass the filename only (in this case the file is assumed to be at `dataset/annotations/`) or the full path.

- **other modifications to train_model.py and eval_model.py**: I modified these scripts to use the datamodule. 

### Pending bits:
- list of files to exclude: I would like to make the list of files to exclude to use wildcards / regex, rather than explicitly passing the names of the files. Will open an issue for it.
- change CLI input arguments for `evaluate_model.py` script. Right now, we are passing directly the list of full paths to the image directories and the list of full paths to annotation files, which is not the most convenient. I think we can factor out and reuse the bits from the `train_model.py` script, but probably better as a separate PR. The README reflects the syntax of the current status.
- tracking the `split_seed` in MLFlow - I think right now it is not being logged correctly (any help here is more than welcome).
- refactor the `eval_model.py` script to use Lightning pose functionalities.

### CODE REVIEW

File: crabs/detection_tracking/evaluate_model.py [hunk 1]
```diff
@@ -150,18 +150,18 @@ def main(args) -> None:
         help="location of trained model",
     )
     parser.add_argument(
-        "--main_dir",
+        "--images_dirs",
```
[the remark this answers is not in the document]
Comment (nikk-nikaznan) [reply crabs/detection_tracking/evaluate_model.py:153]: Ah after read your comment again, I guess it's fine to leave for now. As seems we need to change this file in the future to reflect the `train_model`. I am happy to do this in separate PR after we conclude all data loader related PR.

File: crabs/detection_tracking/train_model.py [hunk 2]
```diff
@@ -32,37 +36,81 @@ class DectectorTrain:
...
+        # Initialise MLflow logger
         mlf_logger = pl.pytorch.loggers.MLFlowLogger(
             run_name=run_name,
-            experiment_name=args.experiment_name,
+            experiment_name=self.experiment_name,
             tracking_uri="file:./ml-runs",
         )
 
         mlf_logger.log_hyperparams(self.config)
+        mlf_logger.log_hyperparams({"split_seed": self.seed_n})
```
[the remark this answers is not in the document]
Comment (nikk-nikaznan) [reply crabs/detection_tracking/train_model.py:113]: It is log under hyperparams, unless you want it to be logged different way?

File: README.md [hunk 3]
```diff
@@ -20,36 +20,53 @@ requires Python 3.9 or 3.10 or 3.11.
...
+Example (using default annotation file and one dataset):
+
+```bash
+python crabs/detection_tracking/train_model.py --dataset_dirs /home/data/dataset1
+```
+
+Example (passing the full path of the annotation file):
+
+```bash
+python crabs/detection_tracking/train_model.py --dataset_dirs /home/data/dataset1 --annotation_files /home/user/annotations/annotations42.json
```
[the remark this answers is not in the document]
Comment (sfmig) [reply README.md:51]: Yeah, by default I am assuming the dataset structure we specified, but I thought for more flexibility we could have (1) the option to change the json file to use under `annotations`, (2) to point to a totally different annotation file by providing the full path.

Happy to remove the second option if you feel it's unnecessary tho!

File: crabs/detection_tracking/datasets.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            annot
+            for annot, slc in zip(
+                dataset["annotations"], slc_annotations_to_exclude
+            )
+            if not slc
+        ]
+
+        # Write to new file under the same location as original annotation file
+        timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+        out_filename = Path(annotation_file).parent / Path(
```
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/datasets.py]: In theory it shouldn't create a file at every epoch, since the dataset should only be created once, then passed to the dataloader, and then the dataloader is used in training, but will have a look! 👀

**Review by samcunliffe — COMMENTED** [review]
~~Are the tests coming later? Because you're putting a lot of untested code in on PR.~~

~~Looking at [what pytorch lightning do in their tests](https://github.com/Lightning-AI/pytorch-lightning/blob/master/tests/tests_pytorch/core/test_datamodules.py), seems not too difficult. Just instantiate our data module, test moisconfiguring it, test correctly configuring it with some small dummy data?~~

~~We don't need to test the pytorch-lightning methods, just the derived stuff where we've overridden them.~~

... sorry I just saw PR's 2 and 3 have the word test in the title 🤦

File: crabs/detection_tracking/datasets.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            Path(annotation_file).stem + "_filt_" + timestamp + ".json"
+        )
+        with open(out_filename, "w") as f:
+            json.dump(dataset, f)
+
+        return str(out_filename)
+
+
+def plot_sample(imgs: list, row_title: Optional[str] = None, **imshow_kwargs):
```
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/datasets.py:140]: yeah I mentioned this in the PR description (quote below). I found it very convenient to quickly sample and visualise the data but not sure here it's its best location.... Happy to remove it or move it to a visualisation module maybe?

> I got a helper plotting function from Pytorch vision examples, to easily plot samples from the data. But maybe this lives better somewhere else?
> ```
> sample = dataset[0]  # tuple with (image, annotations)
> plot_sample([sample])
> ```

File: crabs/detection_tracking/datamodules.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    def train_dataloader(self) -> DataLoader:
+        """Define dataloader for the training set"""
+        return DataLoader(
+            self.train_dataset,
+            batch_size=self.config["batch_size_train"],
+            shuffle=True,  # a shuffled sampler will be constructed
+            num_workers=self.config["num_workers"],
+            collate_fn=self._collate_fn,
+            persistent_workers=True,
+            multiprocessing_context="fork"
```
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/datamodules.py:198]: I just checked this a bit more, I found out I only need it if I select mps/gpu as accelerator in my mac, and have more than 0 workers. I added these constrains to the if-case in `multiprocessing_context` and `persistent_workers`

File: crabs/detection_tracking/datasets.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        -------
+        str
+            path to new annotation file
+        """
+        # Read annotation file as a dataset dict
+        with open(annotation_file, "r") as f:
+            dataset = json.load(f)
+
+        # Determine images to exclude
+        slc_images_to_exclude = [
```
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/datasets.py]: renamed to `select` - it's a list of booleans that are `True` at the indices of the images to exclude

File: crabs/detection_tracking/datasets.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/datasets.py:139]: We agreed this function may come in handy and that it would be nice to have all viz utils together. To get started I moved the `plot_sample` convenience function to visualization module.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfmig) [reply README.md:51]: We agreed we keep this as is for now.

File: crabs/detection_tracking/train_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/train_model.py:113]: I think in hyperparams is fine for now

File: crabs/detection_tracking/datasets.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                for im, slc in zip(dataset["images"], slc_images_to_exclude)
+                if not slc
+            ]
+            dataset["annotations"] = [
+                annot
+                for annot, slc in zip(
+                    dataset["annotations"], slc_annotations_to_exclude
+                )
+                if not slc
+            ]
```
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/datasets.py]: I removed some of the list comprehensions, I think I saved some lines and hopefully it's more readable!

**Review by nikk-nikaznan — APPROVED** [review]
Everything looks good to me. Thank you!
